### PR TITLE
feat(tree2): add more snapshot tests

### PR DESCRIPTION
**Pull request by connorskees**
Add more tree2 snapshot tests and a README. 

Also adds more docs to some things, reflows comments to ~80 characters, and fixes a data corruption bug.

[ADO#5327](https://dev.azure.com/fluidframework/internal/_workitems/edit/5327/) (readme), [ADO#5326](https://dev.azure.com/fluidframework/internal/_workitems/edit/5326/) (expand snapshot test suite), and [ADO#5517](https://dev.azure.com/fluidframework/internal/_workitems/edit/5517/) (avoid double json stringify in snapshot tests)

### CODE REVIEW

**Comment by connorskees** [comment]
^ I did `git pull upstream next` instead of `git pull upstream main` -_-

File: experimental/dds/tree2/src/test/snapshots/testTrees.ts [hunk 1]
```diff
@@ -82,26 +106,162 @@ function generateTreeRecursively(
...
+	{
+		name: string;
+		summary: ISummaryTree;
+	}[]
+> {
+	const testTrees: {
+		only?: boolean;
+		skip?: boolean;
+		name: string;
+		tree: (a: (b: ISharedTree, validationName: string) => Promise<void>) => Promise<void>;
```
Comment (Abe27342) [inline experimental/dds/tree2/src/test/snapshots/testTrees.ts:119]:
```suggestion
		tree: (takeSnapshot: (tree: ISharedTree, name: string) => Promise<void>) => Promise<void>;
```

File: experimental/dds/tree2/src/test/snapshots/testTrees.ts [hunk 2]
```diff
@@ -58,7 +81,8 @@ function generateTreeRecursively(
 			if (height === 1) {
 				const writeCursor = singleTextCursor({
 					type: leafNodeSchema.name,
-					// TODO: these values show up in the snapshot as "[object Object]", which doesn't seem right.
+					// TODO: these values show up in the snapshot as "[object Object]",
```
Comment (Abe27342) [inline experimental/dds/tree2/src/test/snapshots/testTrees.ts:84]: this comment appears to no longer be true, i'd just remove it.

it would be a problem if it were true.

File: experimental/dds/tree2/src/test/snapshots/testTrees.ts [hunk 3]
```diff
@@ -82,26 +106,162 @@ function generateTreeRecursively(
...
+		tree: (a: (b: ISharedTree, validationName: string) => Promise<void>) => Promise<void>;
+	}[] = [
+		{
+			name: "move-across-fields",
+			tree: async (validate) => {
+				const provider = new TestTreeProviderLite(2);
+				const tree1 = provider.trees[0].view;
+				const tree2 = provider.trees[1].view;
+
+				const initialState: JsonableTree = {
```
Comment (Abe27342) [inline experimental/dds/tree2/src/test/snapshots/testTrees.ts:128]: Maybe leave a note here that we're using older editing APIs and not editable-tree-2 because it didn't support cross-field move at the time of writing.

File: experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts [hunk 4]
```diff
@@ -1968,37 +1963,6 @@ describe("SharedTree", () => {
 	});
 });
 
-/**
- * Updates the given `tree` to the given `schema` and inserts `state` as its root.
- */
-// TODO: replace use of this with initialize or schematize, and/or move them out of this file and use viewWithContent
-function initializeTestTree(
```
Comment (Abe27342) [inline experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts:1975]: I'd avoid proliferating this any further: just inline what you need into the test trees where you use it.

File: experimental/dds/tree2/src/test/snapshots/testTrees.ts [hunk 5]
```diff
@@ -82,26 +96,171 @@ function generateTreeRecursively(
...
-// Currently these tests all replicate pre-attachment states.
-// Coverage for other states should be added (including cases with collaboration).
-export function generateTestTrees(): { name: string; tree: () => ISharedTree }[] {
-	return [
+export function generateTestTrees() {
+	const testTrees: {
+		only?: boolean;
+		skip?: boolean;
+		name: string;
+		tree: (takeSnapshot: (tree: ISharedTree, name: string) => Promise<void>) => Promise<void>;
```
Comment (Abe27342) [inline experimental/dds/tree2/src/test/snapshots/testTrees.ts:104]:
```suggestion
		runScenario: (takeSnapshot: (tree: ISharedTree, name: string) => Promise<void>) => Promise<void>;
```

now that this is no longer just producing 'a tree' seems fit to rename it.

Also--I'd update the usages of `validate` (old name) to `takeSnapshot` as well.

File: experimental/dds/tree2/src/test/snapshots/utils.ts [hunk 6]
```diff
@@ -4,20 +4,134 @@
...
+		case SummaryType.Attachment:
+			return "blob";
+		case SummaryType.Tree:
+			return "tree";
+		default:
+			throw new Error(`Unknown type: ${type}`);
+	}
+}
+
+function serializeTree(parentHandle: string, tree: ISummaryTree, rootNodeName: string) {
```
Comment (Abe27342) [inline experimental/dds/tree2/src/test/snapshots/utils.ts:28]:
```suggestion
/**
 * Rather than JSON.stringify the entire `ISummaryTree` for the purpose of writing it to disk,
 * this function more closely resembles how the summary gets serialized to storage.
 * This helps avoid confusion on problems such as double-JSON-stringification.
 */
function serializeTree(parentHandle: string, tree: ISummaryTree, rootNodeName: string) {
```

File: experimental/dds/tree2/src/test/snapshots/summary.spec.ts [hunk 7]
```diff
@@ -4,38 +4,86 @@
...
+// 					it(`for ${name}`, async () => {
+// 						await verifyEqualPastSnapshot(getFilepath(name), summary);
+// 					});
+// 				}
+// 			});
+// 		});
+// 	})
+// 	.catch((e) => {
+// 		throw e;
+// 	});
```
Comment (Abe27342) [inline experimental/dds/tree2/src/test/snapshots/summary.spec.ts:41]:
```suggestion
```

File: packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts [hunk 8]
```diff
@@ -81,7 +81,7 @@ export class OdspSummaryUploadManager {
 		tree: api.ISummaryTree,
 	): Promise<IWriteSummaryResponse> {
 		const containsProtocolTree = isCombinedAppAndProtocolSummary(tree);
-		const { snapshotTree, blobs } = await this.convertSummaryToSnapshotTree(
+		const { snapshotTree, blobs } = this.convertSummaryToSnapshotTree(
```
Comment (Abe27342) [inline packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts:84]: i'd prefer to make this change in a separate PR. Seems fine, but unrelated to tree2 snapshots and if there are any details impacted by JS turn order (I sure hope not) it could be observable.

**Review by Abe27342 — COMMENTED** [review]
looks good minus a couple nits

**Review by Abe27342 — APPROVED** [review]

**Review by Abe27342 — APPROVED** [review]
